Services/Database/classes/Atom/class.ilAtomQuery.php

```diff
@@ -0,0 +1,320 @@
+	}
+
+
+	/**
+	 * @param $isolation_level
+	 * @param $anomaly
+	 * @return bool
+	 * @throws \ilDatabaseException
+	 */
+	public static function isThereRiskThat($isolation_level, $anomaly) {
+		static::checkIsolationLevel($isolation_level);
+		static::checkAnomaly($anomaly);
+
+		return in_array($anomaly, static::getPossibleAnomalies($isolation_level));
+	}
+
+
+	/**
+	 * @param $isolation_level
+	 * @return array
+	 */
+	public static function getPossibleAnomalies($isolation_level) {
+		static::checkIsolationLevel($isolation_level);
+
+		return self::$anomalies_map[$isolation_level];
+	}
+
+
+	/**
+	 * @param $isolation_level
+	 * @throws \ilDatabaseException
+	 */
+	public static function checkIsolationLevel($isolation_level) {
+		// The following Isolations are currently not supported
+		if (in_array($isolation_level, array( self::ISOLATION_READ_UNCOMMITED, self::ISOLATION_READ_COMMITED, self::ISOLATION_REPEATED_READ ))) {
+			throw new ilDatabaseException('This isolation-level is currently unsupported');
+		}
+		// Check if a available Isolation level is selected
+		if (!in_array($isolation_level, self::$available_isolations_levels)) {
+			throw new ilDatabaseException('Isolation-Level not available');
+		}
+	}
+
+
+	/**
+	 * @param $anomalie
+	 * @throws \ilDatabaseException
+	 */
+	public static function checkAnomaly($anomalie) {
+		if (!in_array($anomalie, self::$available_isolations_levels)) {
+			throw new ilDatabaseException('Isolation-Level not available');
+		}
+	}
+
+
+	/**
+	 * @throws \ilDatabaseException
+	 */
+	protected function checkQueries() {
+		foreach ($this->queries as $query) {
+			if (!$query instanceof Closure) {
+				throw new ilDatabaseException('Please provide a Closure with your database-actions by adding with ilAtomQuery->addQueryClosure(function($ilDB) use ($my_vars) { $ilDB->doStuff(); });');
+			}
+		}
+	}
+
+
+	/**
+	 * @return bool
+	 */
+	protected function hasWriteLocks() {
+		$has_write_locks = false;
+		foreach ($this->tables as $table) {
+			$lock_level = $table[1];
+			if ($lock_level == self::LOCK_WRITE) {
+				$has_write_locks = true;
+			}
+		}
+
+		return $has_write_locks;
+	}
+
+
+	/**
+	 * @return array
+	 */
+	protected function getLocksForDBInstance() {
+		$locks = array();
+		foreach ($this->tables as $table) {
+			$table_name = $table[0];
+			$lock_level = $table[1];
+			$locks[] = array( 'name' => $table_name, 'type' => $lock_level );
+		}
+
+		return $locks;
+	}
+
+
+	/**
+	 * @throws ilDatabaseException
+	 */
+	protected function runQueries() {
+		foreach ($this->queries as $query) {
+			/**
+			 * @var $query Closure
+			 */
+			$query($this->ilDBInstance);
+		}
+	}
+
+
+	/**
+	 * @throws \ilDatabaseException
+	 */
+	protected function runWithTransactions() {
+		$e = null;
+		$i = 0;
+		do {
+			try {
+				$this->ilDBInstance->beginTransaction();
+				$this->runQueries();
+				$this->ilDBInstance->commit();
+			} catch (ilDatabaseException $e) {
+				if ($i > 10) {
+					throw $e;
```

Comment:
I did not read the details of the code yet. Why does the transaction code do multiple attempts and the locking code not? What kind of exeptions do we expect in the first case? Are deadlock situations somethink we risk in the second case?